dvc: cache graph/stages/pipelines

efiop opened the pull request:
Signed-off-by: Ruslan Kuprieiev <ruslan@iterative.ai>

Related to #755 

* [x] Have you followed the guidelines in our
      [Contributing document](https://dvc.org/doc/user-guide/contributing)?

* [x] Does your PR affect documented changes or does it add new functionality
      that should be documented? If yes, have you created a PR for
      [dvc.org](https://github.com/iterative/dvc.org) documenting it or at
      least opened an issue for it? If so, please add a link to it.

-----

Suor commented:
How is this related to #755?

On patch itself, it looks so fragile that it doesn't look like it caching anything - it's too easy call `.reload()` to burn the cache, it's probably there already. `.reload()` itself doesn't make sense - it precalculates everything and then deletes it, so that it would be calculated on access again.

This could be done the same as here, but:
- `.reset()` instead if `.reload()`, only dels things
- `.reset()` is called after `.tree` assignment or other tree changes
- cached properties are used as usual, not precalculated
- consider replacing some or most cached properties with simple properties

Also, you are calculating graphs twice because how `.graph` and `.active_graph` are implemented, you can do it in one pass:
```python
@cached_property
def graph(self):
    self.graph, self.active_graph = self._graph()
    return self.graph
```

Another thing is you resurrected concepts of active stages and active pipelines, which we dropped some time ago and altered `.reproduce()` behaviour or implementation - I don't know what it does now, but let's not mix this change with cache addition.

efiop commented:
> How is this related to #755?

I need to implement a stage locker that needs to know what DAG looks like in order to lock dependencies and stuff, and this patch helps avoid passing `self.stages()` everywhere.

> On patch itself, it looks so fragile that it doesn't look like it caching anything - it's too easy call .reload() to burn the cache, it's probably there already. .reload() itself doesn't make sense - it precalculates everything and then deletes it, so that it would be calculated on access again.

Yeah, `reload`'s purpose is to burn the cache, correct, what is wrong with that?  It pre-calculates as a workaround to attribute error when you are trying to `del` some cached property to invalidate it, when it hasn't been called previously. So pre-calculation happens only when cache is empty. Would love to hear how to `del` empty `funcy.cached_property` in the correct way :slightly_smiling_face:

> This could be done the same as here, but:
> .reset() instead if .reload(), only dels things
> .reset() is called after .tree assignment or other tree changes
> cached properties are used as usual, not precalculated
> consider replacing some or most cached properties with simple properties

That is precicesly what reload is meant to do, I've just used cached_property, but should've used good old read-only `property` and `_property` in it. `reset` indeed sounds better than `reload`, thanks, I'll rename it.

> Also, you are calculating graphs twice because how .graph and .active_graph are implemented, you can do it in one pass:

Yep, I should really start adding tickboxes to the PR description to list the issues I'm aware of and that I will fix before WIP becomes a final patch.

> Another thing is you resurrected concepts of active stages and active pipelines, which we dropped some time ago and altered .reproduce() behaviour or implementation - I don't know what it does now, but let's not mix this change with cache addition.

It was an incorrect decision to drop it everywhere. It is not needed for pull/push/etc, but is def needed for repro, so I've fixed that.

Suor commented:
To remove attribute/cached prop, which you are not sure is there, you can:
```python
self.__dict__.pop('graph', None)
```
Maybe I should consider adding silent deletion of `cached_property` to funcy. Swithing to simple properties will also work, but will be a bit more verbose:

```python
@property
def graph(self):
    if not getattr(self, '_graph', None):
        self._graph, self._active_graph = self._build_graph()
    return self._graph
```

On active things, `active_stages` are not used at all, other things are used only in repro, so they should go there. Also, there is no need to cache everything, once graph is built querying it should be much faster - at least that doesn't require fs access. So instead of cached props on `Repo` simple vars in `reproduce()` should work.

As, you can see we are discussing two things here in parallel, which complicate each other. So I suggest moving repro refactor to a separate PR.

Suor commented:
BTW, I still don't understand why you are calling `.reload()` so often, cache should be invalidated on change only.

efiop commented:
> BTW, I still don't understand why you are calling .reload() so often, cache should be invalidated on change only.

@Suor Cache should be reloaded after repo lock is taken, as someone might've deleted/added/modified stages, while we didn't have the lock. And it needs to reloaded whenever we change `self.tree`.

Suor commented:
@efiop So that should be synchronized in code somehow, simply repeating `self.reload()` everywhere is error prone. We also should protect from graph collection, while lock is not hold.

Summarizing: 
- check if we hold the lock before graph collection (need a fast way to check that, i.e. add `is_locked` prop on the lock)
- drop cache on lock release (if it's accessed after that it could be stale already)

Suor commented:
A side note: looks like releasing lock on runs, may cost us some performance - graph needs to be rebuilt if accessed.

efiop commented:
Btw, this also solves https://github.com/iterative/dvc/pull/2519#issuecomment-533757333 . And I've found a few `@locked` that I've missed during https://github.com/iterative/dvc/pull/2519.

Suor commented:
It's very hard to review while two things, caching and reproduce refactor, are intermixed here. Would you mind separating these?

efiop commented:
@Suor it is not trivial to decouple, as they are closely tied together.

efiop commented:
@Suor Or are you specifically about `elif pipeline or all_pipelines` and the fact that i'm using get_active_pipeline there? In that case, that one is a trivial quickfix.

efiop commented:
> I think we should remove all that "active" stuff. This has no sense, especially in Repo context, if some of that is needed in reproduce then it can be calculated there.

We shouldn't remove it, as it is a core feature for `repro` and the only part where it was broken is for `--all-pipelines` and `--pipeline` options, where it was using `self._get_pipeline` instead of using an active pipeline. In regular `dvc repro` cases without that option, `dvc repro` was and is using active graph.

efiop commented:
> then it can be calculated there.

graph and active graph are collected at the same time, so it makes sense to keep `self.active_graph` as a symmetry for `self.graph`.

efiop commented:
> So there will be disagreement, while cache stuff might be merged much faster.

@Suor If your only concern is https://github.com/iterative/dvc/pull/2529#issuecomment-534536226 , I could remove that if it would hold us off.

shcheklein commented on dvc/repo/__init__.py:
```diff
@@ -360,14 +391,38 @@ def graph(self, stages=None, from_directory=None):
+
+    @property
+    def graph(self):
+        return self._graph[0]
```
why is graph[0] something special?

shcheklein replied:
could we use names tuple or something to make it explicit that 1 is active, 0 is not. Otherwise it's easy to confuse with the first DAG, first pipeline, etc

shcheklein commented on dvc/repo/add.py:
```diff
@@ -34,12 +36,19 @@ def add(repo, target, recursive=False, no_commit=False, fname=None):
+    with repo.state:
+        stages = _create_stages(repo, targets, fname)
+
+        repo.check_dag(stages)
```
probably, not related to this PR - would be great if add can do add modify and check graph dynamically, w/o recalculating it again on every add ... what about -R, btw - will it force recalculation on every single file in the directory?

efiop replied:
So for `-R` it won't recalculate it multiple times, as `_find_all_targets(repo, target, recursive) ` above will handle all of that at the same time. That being said, you have a good point about `check_dag` not really needing to re-collect graph in these cases, I'll double-check that.

shcheklein commented on dvc/repo/__init__.py:
```diff
@@ -155,7 +157,8 @@ def _ignore(self):
 
     def check_dag(self, stages):
```
can we actually change names a bit - stages -> new_stages or check_modified_dag?

shcheklein replied:
it looks like it's used in a few other places and should be updated?

shcheklein commented on dvc/repo/__init__.py:
```diff
@@ -155,7 +157,8 @@ def _ignore(self):
     def check_dag(self, stages):
         """Generate graph including the new stage to check for errors"""
-        self.graph(stages=stages)
+        self.reset()
```
also, why do we need to reset it here? it probably means that it's not just a regular check, but actual update or recalculate graph. Is there a problem that we will have some discrepancy between the in-memory representation and what we have on disk for some while? What I want probably to ask is this - if our intention is to check then should we avoid changing the cached properties and use reset when it's actually need?

efiop commented:
@Suor @shcheklein You are right about `active` stuff. I've refactored it so we take original G and cut-off locked deps to achieve active graph and now we are only doing and using that in reproduce. Looks much better know IMHO.

Suor commented on dvc/repo/add.py:
```diff
@@ -56,21 +65,15 @@ def _find_all_targets(repo, target, recursive):
-
-            if not no_commit:
-                stage.commit()
+        if not stage:
```
Is this possible? Or is this some code artifact?

efiop replied:
Possible if already cached.

Suor commented on dvc/repo/graph.py:
```diff
@@ -0,0 +1,29 @@
+    import networkx as nx
+    from dvc.exceptions import CyclicGraphError
+
+    cycles = list(nx.simple_cycles(graph))
```
Only need to get first, with either:
```python
from funcy import first

cycle = first(nx.simple_cycles(graph))
```
or
```
cycle = next(nx.simple_cycles(graph), None)
```

Suor commented on dvc/repo/reproduce.py:
```diff
@@ -32,6 +32,23 @@ def _reproduce_stage(stages, node, **kwargs):
+
+    active = G.copy()
+    stages = nx.get_node_attributes(G, "stage")
+    for node in G.nodes():
```
As I can see you don't need to call `.nodes()`, it's possible to iterate graph directly and there is no need to prepare attrs you may use `stage = G.nodes[node]['stage']`. Or iterate `G.nodes.items()`

BTW, why don't we add stages to graph directly? Then it would be simply:

```python
for stage in G:
    if not stage.locked:
        continue
    # ...
```

efiop replied:
Addressed the `nodes()` issue.

> BTW, why don't we add stages to graph directly? Then it would be simply:

Mostly a legacy I think. Might take a look at some point.

Suor commented on dvc/repo/reproduce.py:
```diff
@@ -32,6 +32,23 @@ def _reproduce_stage(stages, node, **kwargs):
+        for n in nx.dfs_postorder_nodes(G, node):
+            if n == node:
+                continue
+            if n in active.nodes():
```
```suggestion
            if n in active:
```

Suor commented on dvc/repo/reproduce.py:
```diff
@@ -56,43 +73,38 @@ def reproduce(
             for node in G.nodes():
                 if G.in_degree(node) == 0:
-                    targets.append(os.path.join(self.root_dir, node))
+                    targets.append(attrs[node])
```
This all could be:
```python
targets = [stage for G in pipelines for stage in G if G.in_degree(stage) == 0]
```

efiop replied:
That is terrible to do such a long comprehensions. Much better to just leave them structured.

Suor replied:
You can break list comprehension into several lines, this would be the same but more efficient and descriptive. Collecting results with `list.append()` always looked like a smell to me.

efiop commented on dvc/repo/tag/list.py:
```diff
@@ -1,3 +1,7 @@
+from dvc.repo import locked
+
+
+@locked
```
It it supposed to reset it because it no longer guarantees that it will stay intact. But `locked` here is about locking the dvc repository, so that no other dvc instance can add/remove dvcfiles from the project, which would affect `self.collect` used below.

pared replied:
Ok, thanks for the clarification, I only looked at `locked` as a mechanism caching the graph. It makes sense.

efiop commented on dvc/repo/__init__.py:
```diff
@@ -153,42 +169,42 @@ def _ignore(self):
+                stage = attrs[node]
+                if stage.path.startswith(target + os.sep):
+                    ret.append(stage)
+            return ret
```
The order matters. E.g. for `repro --recursive --single-item` we need to run those in the proper order. Might be better to do that not in `collect`, but rather in `reproduce` itself. Need to take a look separately.

Suor replied:
So this one can also be significantly simplified by using stages as nodes.

efiop commented on dvc/repo/__init__.py:
```diff
@@ -153,42 +169,42 @@ def _ignore(self):
+            return ret
 
         stage = Stage.load(self, target)
         if not with_deps:
```
It is more a bug though. `--with-deps --recursive` is a perfectly valid combination. I'll adjust it with a test later.

Suor replied:
So this goes to separate issue?

efiop replied:
@Suor Yes.

efiop commented:
> So the big things are done. Except one - check that graph is not collected while lock is not hold. That will protect us from forgetting @locked basically. This could be done separately though.

Definitely separately. Also we do collect graph without the lock on API/import/get calls when dealing with external repo. These days we can bring that back for the symmetry, as we have universal locks now and are not afraid of NFS and friends.

Suor commented on dvc/repo/__init__.py:
```diff
@@ -21,14 +20,20 @@
         with repo.lock:
-            return f(repo, *args, **kwargs)
+            ret = f(repo, *args, **kwargs)
+            # Our graph cache is no longer valid after we release the repo.lock
```
Why it's not valid? I would write that concurrent dvc process might change stage files and this graph cache might be stale.
